examples/cifar10/train.cpp

```diff
@@ -104,21 +119,58 @@ void train_cifar10(string data_dir_path, double learning_rate, ostream &log) {
                           n_train_epochs, on_enumerate_minibatch,
                           on_enumerate_epoch);
 
-  cout << "end training." << endl;
+  std::cout << "end training." << std::endl;
 
   // test and show results
-  nn.test(test_images, test_labels).print_detail(cout);
-
+  nn.test(test_images, test_labels).print_detail(std::cout);
   // save networks
-  ofstream ofs("cifar-weights");
+  std::ofstream ofs("cifar-weights");
   ofs << nn;
 }
 
+static core::backend_t parse_backend_name(const char *name) {
+  static const char *names[] = {
+    "internal", "nnpack", "libdnn", "avx", "opencl",
+  };
+  for (size_t i = 0; i < sizeof(names) / sizeof(names[0]); ++i) {
+    if (strcasecmp(name, names[i]) == 0) {
+      return (core::backend_t)i;
+    }
+  }
+  return core::default_engine();
+}
+
 int main(int argc, char **argv) {
-  if (argc != 3) {
-    cerr << "Usage : " << argv[0] << " path_to_data (example:../data)"
-         << " learning rate (example:0.01)" << endl;
+  double learning_rate         = 0.1;
+  int epochs                   = 30;
+  std::string data_path        = "";
+  int minibatch_size           = 10;
+  core::backend_t backend_type = core::default_engine();
+  for (int count = 1; count + 1 < argc; count += 2) {
+    std::string argname(argv[count]);
+    if (argname == "--learning_rate")
+      learning_rate = atof(argv[count + 1]);
+    else if (argname == "--epochs")
+      epochs = atoi(argv[count + 1]);
+    else if (argname == "--minibatch_size")
+      minibatch_size = atoi(argv[count + 1]);
+    else if (argname == "--backend_type")
+      backend_type = parse_backend_name(argv[count + 1]);
+    else if (argname == "--data_path")
+      data_path = std::string(argv[count + 1]);
+    else
+      std::cout << "argument " << argname << " isn't supported.";
+  }
+  if (data_path == "") {
+    std::cerr << "Data path not specified. Example of usage :\n"
+              << argv[0]
+              << " --data_path ../data --learning_rate 0.01 --epochs 30 "
+              << "--minibatch_size 10 --backend_type internal" << std::endl;
     return -1;
   }
-  train_cifar10(argv[1], stod(argv[2]), cout);
-}
+  std::cout << "Running with learning rate " << learning_rate
+            << " and minibatch of size " << minibatch_size << " for " << epochs
+            << " epochs." << std::endl;
```

Comment:
I personally prefer hyperparameters to be printed on separate lines so that they can easily parsed by programs.